### ISSUE TITLE
Running several tests simultaneously on the same machine can lead to errors

### ISSUE THREAD

**Issue by artikz** [xiaocong/uiautomator#209]
For example, let's have the following sequence of statements executed in the given order in two different processes:

Process 1: device = uiautomator.Device(serial="1")
Process 2: device = uiautomator.Device(serial="2")
Process 1: device.screen.on()
Process 2: device.screen.on()

After this both processes will be attached to the same device "2", because local_port calculated in AutomatorServer.__init__() is not used until AutomatorServer.start(), and there is no way for process 2 to know that the port is already used when it selects it.

In addition adb forward will silently rebind port by default without returning an error, so neither of the processes will know that it talks to incorrect device.

**Comment by artikz** [comment]
I've sent a pull request that should fix this issue: https://github.com/xiaocong/uiautomator/pull/208

**Comment by codeskyblue** [comment]
Just tested, everything works fine.

**Comment by artikz** [comment]
Do you mean that you can't reproduce the problem?

**Comment by codeskyblue** [comment]
yes

**Comment by artikz** [comment]
This is how to reproduce it:

1. connect 2 devices to a computer

2. copy the scripts and put correct device's serials in placeholders:

test1.py: https://gist.github.com/artikz/c4e9d87820f9dad2a4bacab11a45692b
test2.py: https://gist.github.com/artikz/af3d7feb4706d7b7c8203fec0b5a900a

They just initialize uiautomator.Device() and take screenshot in 60 seconds after initialization.

3. make sure that you don't have any adb port forwarding set up (if you have, then it will trigger another flow in uiautomator code and will work correctly):

$ adb forward --remove-all

4. start test1.py in one terminal and after a few seconds start test2.py in another terminal

5. see that they take screenshot of the same device, also check output of:

$ adb forward --list

You'll see that there is only one forward set up.

**Comment by codeskyblue** [comment]
Maybe you can try my fork version <https://github.com/openatx/atx-uiautomator> but I remember I did not change the code related to `adb forward`.

**Comment by SivaKranthiKumar** [comment]
We also faced similar issue while executing on 2 devices simultaneously

**Comment by artikz** [comment]
@codeskyblue, reproduced it with your fork as well

**Comment by codeskyblue** [comment]
@artikz finally, I meet the same problem, like you. try to fix it tomorrow.

**Comment by artikz** [comment]
The fix is here: https://github.com/xiaocong/uiautomator/pull/208

**Comment by codeskyblue** [comment]
I see that pr, very good code with lot of test codes, I just need a little change in order to use in [atx-uiautomator](https://github.com/openatx/atx-uiautomator)

**Comment by SivaKranthiKumar** [comment]
Still we are observing some issues in CI executions even after using @artikz fix. Need to debug

**Comment by artikz** [comment]
@SivaKranthiKumar, what issues do you see? Same as here? (I.e., test talks to incorrect device)

**Comment by SivaKranthiKumar** [comment]
Yeah. Other device operations were happening in some other device. But didn't debug be them yet. We did apply your fix

**Comment by artikz** [comment]
This is interesting. If you will be able to understand why it didn't work, please, let me know
